### PR TITLE
Fix library instrumented tests

### PR DESCRIPTION
**Pull request by tatocaster**
- Fixed test in `ContactStoreDSLKtTest.kt`, matched fields
- Edited `assertContactUpdated` - compare object contents

### CODE REVIEW

File: library/src/androidTest/java/com/alexstyl/contactstore/ContactStoreTestBase.kt [hunk 1]
```diff
@@ -25,7 +25,7 @@ abstract class ContactStoreTestBase {
         val actual = store.fetchContacts(
             predicate = ContactPredicate.ContactLookup(inContactIds = listOf(editedContact.contactId)),
             columnsToFetch = editedContact.columns
-        ).first().first()
+        ).first().first().mutableCopy()
```
[the remark this answers is not in the document]
Comment (alexstyl) [reply library/src/androidTest/java/com/alexstyl/contactstore/ContactStoreTestBase.kt:28]: Equality between `PartialContact`s and `MutableContact`s does not depend on the type of the class. the `mutableCopy()` should not be needed and the issue is probably somewhere else

**Review by alexstyl — COMMENTED** [review]
Thank you for the contribution again @tatocaster. It's a bit embarasing to have tests failing as I probably I forgot to run the tests before pushing some changes it seems. 

I am currently not on my laptop and I can properly check the PR later in the day (if not tomorrow morning), as I need to see what exactly is failing to say if this is the right solution or not.

It would help if you could update the PR with the log of what is failing. It's totally fine if it is too much hussle.

**Comment by tatocaster** [comment]
This is the output from `ContactStoreDSLKtTest.kt`
```
Expected :SaveRequest(_requests=[Insert(contact=MutableContact(contactId=-1, displayName='pref Paolo Mid Melendez, Suf' lookupKey=null, isStarred=false, columns=[com.alexstyl.contactstore.ContactColumn$Phones@af765e6, com.alexstyl.contactstore.ContactColumn$Mails@18ec2d30, com.alexstyl.contactstore.ContactColumn$Note@38c3ba52, com.alexstyl.contactstore.ContactColumn$Events@16291182, com.alexstyl.contactstore.ContactColumn$PostalAddresses@27c09ef4, com.alexstyl.contactstore.ContactColumn$Image@5e3cb291, com.alexstyl.contactstore.ContactColumn$Names@7b88822e, com.alexstyl.contactstore.ContactColumn$Nickname@257339dc, com.alexstyl.contactstore.ContactColumn$WebAddresses@66724b8, com.alexstyl.contactstore.ContactColumn$Organization@3d74a6df, com.alexstyl.contactstore.ContactColumn$GroupMemberships@252caad, com.alexstyl.contactstore.ContactColumn$ImAddresses@3ed3bee2, com.alexstyl.contactstore.ContactColumn$Relations@12e0f64, com.alexstyl.contactstore.ContactColumn$SipAddresses@19f413c0], imageData=ImageData(raw=[105, 109, 97, 103, 101, 32, 100, 97, 116, 97]), phones=[LabeledValue(value=PhoneNumber(raw=555-11-23), label=com.alexstyl.contactstore.Label$LocationHome@2b18ee27, id=null)], mails=[LabeledValue(value=MailAddress(raw=paolo@paolo.com), label=com.alexstyl.contactstore.Label$LocationWork@64654fd5, id=null)], events=[LabeledValue(value=EventDate(dayOfMonth=3, month=12, year=2021), label=com.alexstyl.contactstore.Label$DateBirthday@4c53c497, id=null)], postalAddresses=[LabeledValue(value=PostalAddress(street=85 Somewhere Str, poBox=, neighborhood=, city=, region=, postCode=, country=), label=com.alexstyl.contactstore.Label$LocationHome@2b18ee27, id=null)], webAddresses=[LabeledValue(value=WebAddress(raw=www.paolo.com), label=com.alexstyl.contactstore.Label$LocationWork@64654fd5, id=null)], imAddresses=[], sipAddresses=[], linkedAccountValues=N/A, note=Note(raw=Note to self), groups=[GroupMembership(groupId=123, _id=null)], relations=[], organization=Org, jobTitle=Job Title, firstName=Paolo, lastName=Melendez, middleName=Mid, prefix=pref, suffix=Suf, phoneticLastName=null, phoneticFirstName=null, phoneticMiddleName=null, fullNameStyle=0, phoneticNameStyle=0, nickname=null))])
Actual   :SaveRequest(_requests=[Insert(contact=MutableContact(contactId=-1, displayName='pref Paolo Mid Melendez, Suf' lookupKey=null, isStarred=false, columns=[com.alexstyl.contactstore.ContactColumn$Phones@af765e6, com.alexstyl.contactstore.ContactColumn$Mails@18ec2d30, com.alexstyl.contactstore.ContactColumn$Note@38c3ba52, com.alexstyl.contactstore.ContactColumn$Events@16291182, com.alexstyl.contactstore.ContactColumn$PostalAddresses@27c09ef4, com.alexstyl.contactstore.ContactColumn$Image@5e3cb291, com.alexstyl.contactstore.ContactColumn$Names@7b88822e, com.alexstyl.contactstore.ContactColumn$Nickname@257339dc, com.alexstyl.contactstore.ContactColumn$WebAddresses@66724b8, com.alexstyl.contactstore.ContactColumn$Organization@3d74a6df, com.alexstyl.contactstore.ContactColumn$GroupMemberships@252caad, com.alexstyl.contactstore.ContactColumn$ImAddresses@3ed3bee2, com.alexstyl.contactstore.ContactColumn$Relations@12e0f64, com.alexstyl.contactstore.ContactColumn$SipAddresses@19f413c0], imageData=ImageData(raw=[105, 109, 97, 103, 101, 32, 100, 97, 116, 97]), phones=[LabeledValue(value=PhoneNumber(raw=555-11-23), label=com.alexstyl.contactstore.Label$LocationHome@2b18ee27, id=null)], mails=[LabeledValue(value=MailAddress(raw=paolo@paolo.com), label=com.alexstyl.contactstore.Label$LocationWork@64654fd5, id=null)], events=[LabeledValue(value=EventDate(dayOfMonth=3, month=12, year=2021), label=com.alexstyl.contactstore.Label$DateBirthday@4c53c497, id=null)], postalAddresses=[LabeledValue(value=PostalAddress(street=85 Somewhere Str, poBox=, neighborhood=, city=, region=, postCode=, country=), label=com.alexstyl.contactstore.Label$LocationHome@2b18ee27, id=null)], webAddresses=[LabeledValue(value=WebAddress(raw=www.paolo.com), label=com.alexstyl.contactstore.Label$LocationWork@64654fd5, id=null)], imAddresses=[], sipAddresses=[], linkedAccountValues=N/A, note=Note(raw=Note to self), groups=[GroupMembership(groupId=123, _id=null)], relations=[], organization=Org, jobTitle=Job Title, firstName=Paolo, lastName=Melendez, middleName=Mid, prefix=pref, suffix=Suf, phoneticLastName=, phoneticFirstName=, phoneticMiddleName=, fullNameStyle=0, phoneticNameStyle=0, nickname=))])
<Click to see difference>

org.junit.ComparisonFailure: expected:<...f, phoneticLastName=[null, phoneticFirstName=null, phoneticMiddleName=null, fullNameStyle=0, phoneticNameStyle=0, nickname=null]))])> but was:<...f, phoneticLastName=[, phoneticFirstName=, phoneticMiddleName=, fullNameStyle=0, phoneticNameStyle=0, nickname=]))])>
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at com.alexstyl.contactstore.ContactStoreDSLKtTest$insert$1.invokeSuspend(ContactStoreDSLKtTest.kt:55)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at com.alexstyl.contactstore.ContactStoreDSLKtTest.insert(ContactStoreDSLKtTest.kt:12)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:121)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by tatocaster** [comment]
This is the output from `AddValuesToExistingContactContactStoreTest.kt`.
But fails for all the tests in `UpdatesValuesOfExistingContactContactStoreTest.kt` because they all depend on `assertContactUpdated` function

```
java.lang.AssertionError: 
Expected: <MutableContact(contactId=762, displayName='A. Paolo M. Melendez, Z.' lookupKey=null, isStarred=false, columns=[com.alexstyl.contactstore.ContactColumn$Names@ceb2f63], imageData=N/A, phones=N/A, mails=N/A, events=N/A, postalAddresses=N/A, webAddresses=N/A, imAddresses=N/A, sipAddresses=N/A, linkedAccountValues=N/A, note=N/A, groups=N/A, relations=N/A, organization=N/A, jobTitle=N/A, firstName=Paolo, lastName=Melendez, middleName=M., prefix=A., suffix=Z., phoneticLastName=, phoneticFirstName=, phoneticMiddleName=, fullNameStyle=0, phoneticNameStyle=0, nickname=N/A)>
     but: was <PartialContact(contactId=762, displayName='A. Paolo M. Melendez, Z.' lookupKey=LookupKey(value=0r762-294729453F454141313F31432F315B5B), isStarred=false, columns=[com.alexstyl.contactstore.ContactColumn$Names@ceb2f63], imageData=N/A, phones=N/A, mails=N/A, events=N/A, postalAddresses=N/A, webAddresses=N/A, imAddresses=N/A, sipAddresses=N/A, linkedAccountValues=N/A, note=N/A, groups=N/A, relations=N/A, organization=N/A, jobTitle=N/A, firstName=Paolo, lastName=Melendez, middleName=M., prefix=A., suffix=Z., phoneticLastName=, phoneticFirstName=, phoneticMiddleName=, fullNameStyle=1, phoneticNameStyle=0, nickname=N/A)>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at com.alexstyl.contactstore.ContactStoreTestBase.assertContactUpdated(ContactStoreTestBase.kt:30)
	at com.alexstyl.contactstore.ContactStoreTestBase$assertContactUpdated$1.invokeSuspend(Unknown Source:15)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
	at com.alexstyl.contactstore.AddValuesToExistingContactContactStoreTest.updatesNames(AddValuesToExistingContactContactStoreTest.kt:23)
```

**Comment by tatocaster** [comment]
The difference between objects are: `lookupKey` is null for the actual contact and `fullNameStyle` is 0 instead of 1.
mutableCopy fixes that because it modifies some of the properties

**Comment by alexstyl** [comment]
> The difference between objects are: lookupKey is null for the actual contact and fullNameStyle is 0 instead of 1.
mutableCopy fixes that because it modifies some of the properties

@tatocaster Seems like that is the cause of the failure. The `Contact.mutableCopy()` does not keep the lookupKey and the `fullNameStyle` is using the `phoneticNameStyle` instead of the `fullNameStyle` of the original contact.

PS: I had a quick look on the web and noticed that the 'lookupKey' [is not](https://stackoverflow.com/questions/56464107/android-contacts-api-does-lookup-key-still-change-in-android-5-1) as [permanent](https://developer.android.com/guide/topics/providers/contacts-provider#ContactBasics) as the documentation says. Could you update `assertContactUpdated` to check the contents of the `actual` instead of equality? Otherwise it seems like the test will fail because of lookupkey changes. There is a matcher called `equalContents()` that you can use in this case?

**Comment by tatocaster** [comment]
> fullNameStyle is using the phoneticNameStyle instead of the fullNameStyle of the original contact.

Should we leave it like that or?

> Regarding the `ContactStoreDSLKtTest` I noticed that the `MutableContactBuilder`'s default values do not match with `MutableContact`'s default values (i.e. MutableContact's phoneticFirstName defaults to null, but Builder's one defaults to an empty string). 

Yeah, I noticed them too. Thought it was part of the design. I was thinking the same and I wanted your review. Will update asap, thanks

**Comment by alexstyl** [comment]
> Should we leave it like that or?

Let's change it so that the new MutableContact takes the `fullnameStyle` of the previous one. It was probably a copy paste mistake. :)

**Comment by alexstyl** [comment]
@tatocaster heads up: I have Ignored the failing tests on main.

**Comment by tatocaster** [comment]
All tests are passing! ✅  
I'd like to add Github Actions for the future PRs to run all the tests automatically. The idea for another contribution 😃

**Comment by alexstyl** [comment]
Awesome. Seems 💯 

Could you please squash your first 3 commits? I do not think there is a need to have those in git history as the real fix is the commits you pushed today.

After that I will merge to main 🚀 

PS: Having the tests run via Github Actions would be amazing! I haven't used Actions at all, so I would be really keen to see how it works

**Comment by tatocaster** [comment]
@alexstyl  Done! 
Android studio optimized imports, if we are OK with that then it's ready to merge. Otherwise, I can introduce Detekt or Ktlint to the project to avoid such changes.
Let me know what you think

**Comment by alexstyl** [comment]
@tatocaster That is alright for now. The scope of the PR was to fix the tests and you did. 👏 

The imports and coding style is something for an other PR. I will be creating some Github issues today for things to be done. If you are looking for more contributions until then, I think that having unit and Android tests running on PRs would be the most valuable for now.

Merging this to main. Once again thanks a ton for the contribution 💯

**Pull request merged**